### PR TITLE
Add arrow range to SynExpr.Lambda and SynMatchClause

### PR DESCRIPTION
**Pull request by nojaf**

### CODE REVIEW

**Comment by auduchinok** [comment]
@nojaf what is info about particular tokens is needed for, given expression ranges are known and the only token of a kind can easily be found while traversing the token stream in an expression range? It seems a bit wasteful to add every token this way.

**Comment by auduchinok** [comment]
Also, why use ranges instead of positions, given keywords have fixed lengths?

**Comment by nojaf** [comment]
I would need a range on my side anyway and as `rhs parseState` already returns a range it seems a tad unfortunate to have created a range in the parser, store only the StartPos and create ranges again on the consumer side.

**Comment by nojaf** [comment]
@vzarytovskii could I get a review on this, please?

File: src/fsharp/pars.fsy [hunk 1]
```diff
@@ -3938,30 +3938,34 @@ patternAndGuard:
...
 
   | patternAndGuard patternResult BAR error 
      { let pat, guard, patm = $1 
+       let mArrow, resultExpr = $2
        let mLast = rhs parseState 3 
-       let m = unionRanges $2.Range patm
+       let m = unionRanges resultExpr.Range patm
        // silent recovery
-       [SynMatchClause(pat, guard, $2, m, DebugPointForTarget.Yes)], mLast  }
+       [SynMatchClause(pat, guard, (Some mArrow), resultExpr, m, DebugPointForTarget.Yes)], mLast  }
```
Comment (dsyme) [inline src/fsharp/pars.fsy:3960]: Minor nit - no need for parens around `(Some mArrow)`

**Review by dsyme — APPROVED** [review]

**Comment by dsyme** [comment]
Great work

**Pull request merged**